crates/accelerate/src/sabre_swap/mod.rs

```diff
@@ -390,46 +496,119 @@ fn update_route(
     populate_extended_set(extended_set, dag, front_layer, required_predecessors);
 }
 
+fn gen_swap_epilogue(
+    coupling: &DiGraph<(), ()>,
+    mut from_layout: NLayout,
+    to_layout: &NLayout,
+    seed: u64,
+) -> Vec<[usize; 2]> {
+    // Map physical location in from_layout to physical location in to_layout
+    let mapping: HashMap<NodeIndex, NodeIndex> = from_layout
+        .logic_to_phys
+        .iter()
+        .enumerate()
+        .map(|(v, p)| {
+            (
+                NodeIndex::new(*p),
+                NodeIndex::new(to_layout.logic_to_phys[v]),
+            )
+        })
+        .collect();
+
+    let swaps = token_swapper(
+        coupling,
+        mapping,
+        Some(SWAP_EPILOGUE_TRIALS),
+        Some(seed),
+        None,
+    );
+
+    // Convert physical swaps to virtual swaps
+    swaps
+        .into_iter()
+        .map(|(l, r)| {
+            let ret = [
+                from_layout.phys_to_logic[l.index()],
+                from_layout.phys_to_logic[r.index()],
+            ];
+            from_layout.swap_physical(l.index(), r.index());
+            ret
+        })
+        .collect()
+}
+
 /// Search forwards in the `dag` from all the nodes in `to_visit`, adding them to the `gate_order`
 /// or the current `front_layer` as appropriate, and continue inspecting gates until there is
 /// nothing further with no required predecessors.
 ///
 /// The nodes in `to_visit` should all already have no further required predecessors.
-fn route_reachable_nodes(
+fn route_reachable_nodes<F>(
+    seed: u64,
     to_visit: &[NodeIndex],
     dag: &SabreDAG,
     layout: &NLayout,
     coupling: &DiGraph<(), ()>,
     gate_order: &mut Vec<usize>,
     front_layer: &mut FrontLayer,
     required_predecessors: &mut [u32],
-) {
+    node_block_results: &mut HashMap<usize, Vec<BlockResult>>,
+    route_block_dag: &F,
+) where
+    F: Fn(&SabreDAG, NLayout) -> (SabreResult, NLayout),
+{
     let mut to_visit = to_visit.to_vec();
     let mut i = 0;
     // Iterate through `to_visit`, except we often push new nodes onto the end of it.
     while i < to_visit.len() {
         let node = to_visit[i];
         i += 1;
         let (py_node, qubits) = &dag.dag[node];
-        match qubits[..] {
-            [a, b]
-                if !coupling.contains_edge(
-                    NodeIndex::new(layout.logic_to_phys[a]),
-                    NodeIndex::new(layout.logic_to_phys[b]),
-                ) =>
-            {
-                front_layer.insert(node, [a, b]);
+
+        match dag.node_blocks.get(py_node) {
+            Some(blocks) => {
+                // Control flow op. Route all blocks for current layout.
+                let mut block_results: Vec<BlockResult> = Vec::with_capacity(blocks.len());
+                for inner_dag in blocks {
+                    let (inner_dag_routed, inner_final_layout) =
+                        route_block_dag(inner_dag, layout.copy());
```

Comment:
This is where I got tripped up with the mutation of the `layout.copy()` - we allocate an `NLayout` in the `layout.copy()`, but then the return of the call makes it _seem_ to me that there must be an internal allocation within `route_block_dag`.